packages/grafana-data/src/types/variables.ts

```diff
@@ -0,0 +1,93 @@
+import { ComponentType } from 'react';
+import { Observable } from 'rxjs';
+
+import {
+  DataQuery,
+  DataQueryRequest,
+  DataQueryResponse,
+  DataSourceApi,
+  DataSourceJsonData,
+  DataSourceOptionsType,
+  DataSourceQueryType,
+  QueryEditorProps,
+} from './datasource';
+
+/**
+ * Defines new variable support
+ *
+ * Use one of the properties standard | custom | datasource to define the kind of query editor and query support you want to use in the Variable system
+ *
+ * @alpha -- experimental
+ */
+export interface VariableSupport<
+  DSType extends DataSourceApi<TQuery, TOptions>,
+  TQuery extends DataQuery = DataSourceQueryType<DSType>,
+  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>,
+  VariableQuery extends DataQuery = any
+> {
+  /**
+   * Use this property if you want to use the standard query editor supplied in the Variable system
+   */
+  standard?: StandardVariableSupport<DSType, TQuery, TOptions>;
+
+  /**
+   * Use this property if you want to use a custom query and query editor
+   */
+  custom?: CustomVariableSupport<DSType, VariableQuery, TQuery, TOptions>;
```

Comment:
In addition to this do you have to supply the component somewhere?

Reply:
Not sure I understand your question. The interface tells you need to define an editor and a custom query method like so:
```ts
export interface CustomVariableSupport<
  DSType extends DataSourceApi<TQuery, TOptions>,
  VariableQuery extends DataQuery = any,
  TQuery extends DataQuery = DataSourceQueryType<DSType>,
  TOptions extends DataSourceJsonData = DataSourceOptionsType<DSType>
> {
  editor: ComponentType<QueryEditorProps<DSType, TQuery, TOptions, VariableQuery>>;
  query: (request: DataQueryRequest<VariableQuery>) => Observable<DataQueryResponse>;
}
```